thoughts-of-a-rustacean-learning-go.md

```diff
@@ -0,0 +1,244 @@
+
+
+In Go, interfaces get implemented automatically if a type has methods of a matching signature.
+So an interface with no methods is equivalent to `interface{}`; and will be implemented on all types automatically.
+This means that we can't define "marker traits" like in Rust that add a simple layer of type safety over methods.
+It also means that interfaces can only be used to talk of code level behavior, not higher level abstractions.
+For example, in Rust we have the `Eq` trait, which uses the same method as `PartialEq` for equality (`eq(&self, &other)`),
+and the behavior of that method is exactly the same, however the two traits mean fundamentally different things:
+A type implementing PartialEq has a normal equivalence relation,
+whilst one that also implements Eq has a full equivalence relation.
+From the point of view of the code, there's no difference between their behavior.
+But as a programmer, I can now write code that only accepts types with a full equivalence relation, and exploit that guarantee to optimize my code.
+
+Having interfaces be autoimplemented on the basis of the method signature is a rather ergonomic feature in my opinion
+and it reduces boilerplate. However, it's just not what I'm used to and it restricts me from writing certain types of code without using
+dummy methods [like so][dummy] to get static type safety.
+
+
+[dummy]: https://github.com/Manishearth/cs733/blob/e86f22487f53b6bee21415d6b6ca56b8007792cc/assignment3/raft/raft.go#L57
+
+### Packages and imports
+
+Go puts severe restrictions on where I can put my files. All files in a folder are namespaced into the same package
+(if you define multiple packages in one folder it errors out). There's no way to specify portable
+relative paths for importing packages either. To use a package defined in an adjacent folder, I had to do [this][absolute-import],
+whereas in Rust (well, Cargo), it is easy to specify relative paths to packages (crates) [like so][relative-import].
+The import also only worked if I was developing from within my `$GOPATH`, so my code now resides within
+`$GOPATH/src/github.com/Manishearth/cs733/`; and I can't easily work on it elsewhere without pushing and running `go get` everytime.
+
+Rust's module system [does take hints from the file structure], and it can get confusing;
+however the behavior can be nearly arbitrarily overridden if necessary (you can even do [scary things like this][chtulhu-import]).
+
+[absolute-import]: https://github.com/Manishearth/cs733/blob/9842305bab2f8402dbd235c8512a05f57da1cc5f/assignment2/kvstore/server.go#L5
+[relative-import]: https://github.com/servo/servo/blob/c2fc6e311a0cb3ec4c702c77bb5d13f97bd19078/components/script/Cargo.toml#L19
+[chtulhu-import]: https://twitter.com/horse_rust/status/517408273750704128
+[rust-module]: http://doc.rust-lang.org/book/crates-and-modules.html
+
+### Documentation
+
+ Rust's libraries aren't yet well documented; but this is mostly because the libraries are
```

Comment:
> It's totally understandable that Rust's documentation needs work, given that the language's syntax has evolved significantly over the past couple years.

Actually, that's not the issue, many of the libraries are new (in their current form, that is)and unstable. I'll reword it, but I'd like to keep the stability distinction.